### ISSUE TITLE
sql: CREATE FUNCTION ... LANGUAGE SQL

### ISSUE THREAD

**Issue by benesch** [MaterializeInc/materialize#5112]
The goal here is to create support for a fairly restricted subset of PostgreSQL's [`CREATE FUNCTION`](https://www.postgresql.org/docs/current/sql-createfunction.html) command:

```
CREATE FUNCTION add(integer, integer) RETURNS integer
    AS 'select $1 + $2;'
    LANGUAGE SQL
    IMMUTABLE
    RETURNS NULL ON NULL INPUT;
```

To start, only `LANGUAGE SQL` will be supported, since we already have all the code to parse, plan, and execute SQL. Support for other languages (specifically, WASM) is planned, but will require substantially more design work.

Various features of `CREATE FUNCTION` that are explicitly out of scope:

  * `OUT` parameters.
  *  Default arguments.
  * `WINDOW` functions.
  * `STABLE` or `VOLATILE` functions.
  * `SECURITY INVOKER` and `SECURITY DEFINER`.
  * Any of the other parameters that inform the PostgreSQL optimizer about the function's behavior, like `COST` or `ROWS`.

The initial implementation can probably operate like macros/string replacement. Wherever we see a call to a user-defined SQL function in the planner, we can just swap it for the literal text of the function (after substituting parameters). Most of this code already exists in the `sql_op!` macro. As a result I don't anticipate any changes to the decorrelator or dataflow layer will be required.

There is of course a substantial amount of work to be done in the planner and catalog to store UDFs and to resolve them at runtime.

One substantial restriction of this approach is that it will be nearly impossible to support `CREATE OR REPLACE FUNCTION`. I think that's ok. In PostgreSQL, `REPLACE FUNCTION` will actually update existing views to use the new function, while a `DROP` followed by a `CREATE` will not. So we can just not support `OR REPLACE` for now, and we'll be perfectly compatible with PostgreSQL.

**Comment by benesch** [comment]
Oh, just to be clear: these are "scalar" UDFs only. Supporting "table" UDFs will be a simple extension, via the `RETURNS TABLE` syntax. But supporting aggregate UDFs is a long ways away, because aggregates will be tightly coupled to timely/differential.

**Comment by benesch** [comment]
One big blocker to this that occurs to me: name resolution. Our new approach to name resolution pre-resolves all names before planning (#12858). This is going to be quite irritating when we need to support UDFs, because we'll need to support dynamic name resolution for functions. (Right now, because all functions are built in, we can get away with a static lookup table that's hardcoded in the binary.)

What makes name resolution tricky for functions is that you can have overloads: multiple functions with the same name but different types. E.g.:

```
foo(int)
foo(text)
```

These will be two distinct functions with two distinct `GlobalId`s in the catalog. During the name resolution pass, we don't have type information, so we don't have any means of determining _which_ `foo` to resolve.

One option is to hoist type checking into the same pass as name resolution. This was always tentatively the plan. But I'm not sure how cleanly type checking and planning will separate in practice, due to how untyped literals are coerced.

**Comment by sthm** [comment]
I think this is going to be even more relevant in the future when we are seeing more adoption of `WITH MUTUALLY RECURSIVE`. Here's an example I've been working on during a recent hackathon (full source code is [here](https://drive.google.com/file/d/14XvIPF6MU9HqgSbLs77a6owz_RovuXh4/view?usp=sharing)). Note that all three parts of the recursive query share the same where clause.

```sql
\set pi 3.14159265358979323
\set acos_arg sin(p.pickup_latitude*:pi/180)*sin(q.pickup_latitude*:pi/180) + cos(p.pickup_latitude*:pi/180)*cos(q.pickup_latitude*:pi/180)*cos(q.pickup_longitude*:pi/180-p.pickup_longitude*:pi/180)

CREATE VIEW pickup_clusters AS
WITH MUTUALLY RECURSIVE
    core_points(id text, pickup_latitude numeric, pickup_longitude numeric) AS (
        -- find core points
        SELECT p.id, p.pickup_latitude, p.pickup_longitude
        FROM trips p, trips q, demo_params
        WHERE CASE WHEN :acos_arg BETWEEN -1 AND 1 THEN acos(:acos_arg) * 6371000 < demo_params.epsilon ELSE false END
        GROUP BY p.id, p.pickup_latitude, p.pickup_longitude, demo_params.min_points
        HAVING count(DISTINCT (q.pickup_latitude, q.pickup_longitude)) >= demo_params.min_points
    ),
    expand_clusters(id text, pickup_latitude numeric, pickup_longitude numeric) AS (
        -- the initial seed of core_points is kept during fixpoint iteration, but will be filtered out in the result
        SELECT * FROM core_points

        UNION

        -- compute common id for clusters of core points
        SELECT min(q.id), p.pickup_latitude, p.pickup_longitude
        FROM expand_clusters p, expand_clusters q, demo_params
        WHERE CASE WHEN :acos_arg BETWEEN -1 AND 1 THEN acos(:acos_arg) * 6371000 < demo_params.epsilon ELSE false END
        GROUP BY p.id, p.pickup_latitude, p.pickup_longitude
    )
-- select core points and non-core points as the final cluster
SELECT min(q.id) AS cluster_id, p.pickup_latitude, p.pickup_longitude
FROM trips p, expand_clusters q, demo_params
WHERE CASE WHEN :acos_arg BETWEEN -1 AND 1 THEN acos(:acos_arg) * 6371000 < demo_params.epsilon ELSE false END
GROUP BY p.pickup_latitude, p.pickup_longitude
ORDER BY cluster_id
```

I'm using `\set` as a hack to spell out parts of the repeated condition only once. However, this has several drawbacks:
- `\set` appears to remove white spaces, so it cannot include anything that requires white spaces to be parsed properly (like SQL keywords)
- I had to use the same aliases for the tables in all queries (`p` and `q`) although different aliases would have been more natural
- Occurrences of a parameter (like `:acos_arg`) are simply replaced when the query is submitted. If the definition is changed later on, all objects that refer to it need to be dropped and recreated.

Having scalar UDFs as proposed here would help to avoid these drawbacks. Because `WMR` supports implementing more advanced functionality, I think `WMR` would particularly benefit from scalar UDFs.

**Comment by sploiselle** [comment]
Before this work can be done in earnest, I think we need to assign a `GlobalId` to each function signature (i.e. everything with an `Oid` should also have a unique `GlobalId`). I have a very, very, very preliminary sketch of this [in Notion](https://www.notion.so/materialize/Each-function-impl-needs-its-own-GlobalId-a4adccf564e345d4845d7728f821c62d?pvs=4).

**Comment by benesch** [comment]
Thank you, Sean! 🙇🏽
